sled-agent/src/hardware/illumos/mod.rs

```diff
@@ -73,56 +167,194 @@ fn node_name(subsystem_vid: i32, subsystem_id: i32) -> String {
     format!("pci{subsystem_vid:x},{subsystem_id:x}")
 }
 
+fn slot_to_disk_variant(slot: i64) -> Option<DiskVariant> {
+    match slot {
+        0x00..=0x09 => Some(DiskVariant::U2),
+        0x11..=0x12 => Some(DiskVariant::M2),
+        _ => None,
+    }
+}
+
+fn poll_tofino_node(
+    log: &Logger,
+    tofino: &mut TofinoSnapshot,
+    node: &Node<'_>,
+) {
+    if node.node_name() == node_name(TOFINO_SUBSYSTEM_VID, TOFINO_SUBSYSTEM_ID)
+    {
+        tofino.exists = true;
+        tofino.driver_loaded = node.driver_name().as_deref() == Some("tofino");
+        debug!(
+            log,
+            "Found tofino node, with driver {}loaded",
+            if tofino.driver_loaded { "" } else { "not " }
+        );
+    }
+}
+
+fn get_dev_path_of_whole_disk(
+    node: &Node<'_>,
+) -> Result<Option<PathBuf>, Error> {
+    let mut wm = node.minors();
+    while let Some(m) = wm.next().transpose().map_err(Error::DevInfo)? {
+        // "wd" stands for "whole disk"
+        if m.name() != "wd" {
+            continue;
+        }
+        let links = {
+            match DevLinks::new(true) {
+                Ok(links) => links,
+                Err(_) => DevLinks::new(false).map_err(Error::DevInfo)?,
+            }
+        };
+        let devfs_path = m.devfs_path().map_err(Error::DevInfo)?;
+
+        let paths = links
+            .links_for_path(&devfs_path)
+            .map_err(Error::DevInfo)?
+            .into_iter()
+            .filter(|l| {
+                // Devices in "/dev/dsk" have names that denote their purpose,
+                // of the form "controller, disk, slice" or "controller, disk,
+                // partition".
+                //
+                // The suffix of "d0" is typical of an individual disk, and is
+                // the expected device to correspond with the "wd" device in
+                // the "/devices" hierarchy.
+                l.linktype() == DevLinkType::Primary
+                    && l.path()
+                        .file_name()
+                        .map(|f| f.to_string_lossy().ends_with("d0"))
+                        .unwrap_or(false)
+            })
+            .collect::<Vec<_>>();
+
+        if paths.is_empty() {
+            return Err(Error::NoDevLinks(PathBuf::from(devfs_path)));
+        }
+        return Ok(Some(paths[0].path().to_path_buf()));
+    }
+    Ok(None)
+}
+
+// Gather a notion of "device identity".
+//
+// NOTE: What exactly comprises a "reliable, unique identifier" for a disk is
+// still a little bit in flux. While that's being sorted out, we simply produce
+// a string that can uniquely identify the device to Nexus.
+//
+// At the moment, this consists of a "Vendor ID, Serial ID, and model ID".
```

Comment:
I'm open to actually returning these individual fields up to Nexus, rather than as a concatenated string.

However, from Nexus' point-of-view, I'd like to use the ~~Verbal, Somatic, and Material Components~~ Vendor, Serial, and Model IDs as an opaque primary key to identify disks.